src/inatinqperf/adaptors/weaviate_adaptor.py

```diff
@@ -1,257 +1,340 @@
+        vectors = np.vstack([np.asarray(dp.vector, dtype=np.float32) for dp in datapoints])
+        if vectors.ndim != self._EXPECTED_VECTOR_RANK or vectors.shape[1] != self.dim:
+            msg = "DataPoint vectors must be 2-D with shape (n, dim)."
             raise ValueError(msg)
 
-        # Prefer best-effort ingestion: we validate matching lengths above, but leave
-        # strict=False so late mismatches drop extras instead of aborting the batch.
-        for identifier, vector in zip(ids, vectors, strict=False):
-            obj_id = self._make_object_id(int(identifier))
-            # Delete any pre-existing object so POST remains consistent.
-            delete_resp = self._session.delete(
-                f"{self._objects_endpoint}/{self.class_name}/{obj_id}", timeout=self.timeout
-            )
-            if delete_resp.status_code not in {HTTPStatus.OK, HTTPStatus.NO_CONTENT, HTTPStatus.NOT_FOUND}:
-                self._raise_error("failed to delete existing object", delete_resp)
-
-            body = {
-                "id": obj_id,
-                "class": self.class_name,
-                "vector": vector.tolist(),
-                "properties": {"originalId": int(identifier)},
-            }
-            response = self._session.post(self._objects_endpoint, json=body, timeout=self.timeout)
-            if response.status_code not in {200, 201}:
-                self._raise_error("failed to upsert object", response)
-
-    def search(self, q: Query, topk: int, **_: object) -> Sequence[SearchResult]:
-        """Run nearest-neighbor search using GraphQL."""
+        # Ensure the class exists. This is a no-op if already provisioned.
+        self.train_index(vectors)
+
+        for dp, vector in zip(datapoints, vectors, strict=False):
+            obj_id = self._make_object_id(int(dp.id))
+            properties = dict(getattr(dp, "metadata", {}) or {})
+            properties.setdefault(self._ORIGINAL_ID_FIELD, int(dp.id))
+
+            try:
+                # Re-using deterministic object IDs keeps the collection idempotent.
+                self._client.data_object.delete(obj_id, class_name=self.class_name)
+            except Exception as exc:  # pragma: no cover - defensive programming
+                status_code = getattr(exc, "status_code", None)
+                if status_code != HTTPStatus.NOT_FOUND:
+                    self._handle_exception("failed to delete existing object", exc)
+
+            try:
+                self._client.data_object.create(
+                    data_object=properties,
+                    class_name=self.class_name,
+                    uuid=obj_id,
+                    vector=vector.tolist(),
+                )
+            except Exception as exc:  # pragma: no cover - defensive programming
+                self._handle_exception("failed to upsert object", exc)
+
+    def search(
+        self,
+        q: Query | Sequence[float] | np.ndarray,
```

Comment:
This is wrong and breaks the interface. We should only be accepting a single query, there should not be any form of sequence here.

Reply:
Not addressed